### ISSUE TITLE
FC023 can be kinda idiotic sometimes

### ISSUE THREAD

**Issue by lamont-granquist** [acrmp/foodcritic#422]
This code triggers FC023:

``` ruby
unless windows?
  powershell_script "foo" do
    code <<-EOH
    stuff
    EOH
  end
end
```

It is, however, better code than this:

``` ruby
powershell_script "foo" do
  code <<-EOH
  stuff
  EOH
 only_if { windows? }
end
```

In Chef 12.2.0-12.7.0 (at a minimum) the latter code won't actually work if you're not on windows, since the dynamic resource resolver won't be able to lookup powershell_script at all since its provides line is:

``` ruby
      provides :powershell_script, os: "windows"
```

So, if you run the "FC023-approved" codeblock on a (e.g.) solaris box, it'll explode with an exception.

Its also not useful to have a powershell_script resource in the resource collection at all on a solaris box.  You can't ever notify it without it throwing an exception anyway, so its value on the resource collection is zero at best, and negative at worst.

Of course fixing this problem in foodcritic would be pretty hard since it requires semantic information about exactly what kind of guard is being applied around the resource.  The argument could be made that users which hit this should just know to write an exception to FC023.  At the same time if a relatively novice user runs foodcritic they'll get back FC023 and could 'fix' the problem and wind up breaking their cookbook on every O/S other than windows--which is precisely what they're trying to avoid.

We also may reconsider the provides line in the powershell_script resource.  But I think this issue stands even if we wind up working around that, in that the powershell_script resource in the resource collection on a solaris box has exactly zero value.

At the same time, the vastly more common use cases where FC023 gets triggered is probably useful to users.  New users won't necessarily understand that they can't notify a resource if its not put on the resource collection and the proper use of guards is definitely more idiomatic chef.  I don't think it has no value, but it isn't a universal rule that should always be followed with no exceptions.

Perhaps we just need a good documentation update...

**Comment by thommay** [comment]
This is more of a discussion about how chef handles providers, and I started thinking about this when I was writing the `apt_update` provider.
Right now, many resources are abstract - ie, they provide a generic `package` resource, and then we have a concrete `package` implementation per-platform. But for certain resources, like `apt_update` and like `powershell_script`, this isn't the case.
But now, users have to understand that they can't simply have `apt_update` in a recipe and have it be compiled away on a windows box; they have to specifically guard it every time, and they have to deal with notifications properly. We're pushing a lot of complexity that we could *easily* deal with down to our users.

**Comment by lamont-granquist** [comment]
Technically, all we have now for the package resource is a DSL shim that picks out the correct provider.  The base class does not provide the package DSL at all and there's no way to instantiate the base class from the DSL.

I can definitely see the argument that the `apt_update` and `powershell_script` resources should be available everywhere.  I can see the argument for the `apt_update` provider that on the backend it should wire up to a no-op provider.  I'm less sure that `powershell_script` should.

**Comment by lamont-granquist** [comment]
so another issue is that if a node attribute isn't defined and it isn't legal to construct the resource then the two aren't equivalent:

```ruby
# this protects the resource so it never gets created with `must_not_be_nil` set to nil which will fail validation
if node['thingy']
  my_resource "do stuff" do
    must_not_be_nil node['thingy']
  end
end
```

```ruby
# this is going to try to create a resource with `must_not_be_nil` set to nil which will fail validation
  my_resource "do stuff" do
    must_not_be_nil node['thingy']
    only_if { node['thingy'] }
  end
```

**Comment by lamont-granquist** [comment]
closed by #523
